### PR TITLE
Issue 3368: Ensure tier2 for Pravega is configurable when deployed by Kubernetes framework

### LINKED ISSUE #3368: Ensure tier2 for Pravega is configurable when deployed by Kubernetes framework.
**Problem description**
Currently Kubernetes framework only supports `NFS` as tier2.
Enable support for `HDFS` and `extended S3  object store`.

**Problem location**
`kubernetes.AbstractService`

**Suggestions for an improvement**
Enable support for other tier2.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
Enable support for `HDFS` and `extended S3 object store` based tier2 for pravega deployed using Kubernetes test framework.

**Purpose of the change**  
Fixes #3368 

**What the code does**  
It enables support to run system tests to run on  `ECS` and `HDFS` based tier2 in addition to `NFS`. Now the framework can pass the tier2 configuration to the `pravega-operator` as specified at 
https://github.com/pravega/pravega-operator/blob/master/example/cr-detailed.yaml#L61

A new system property `tier2Type` has been added which indicates the type of tier2. It can be either NFS, ECS or HDFS. (By default NFS will be used if this is not set). Tier2 configuration can be specified by using a system property `tier2Config` which is a comma separated key value pairs.

Command to Run tests:
* NFS based tier2: (same as earlier)
```gradle --info startK8SystemTests -DdockerRegistryUrl=private-repo:8116 -DimagePrefix=prefix -DimageVersion=0.5.0-2111.f48a00a ```

* HDFS based tier2
```gradle --info startK8SystemTests -DdockerRegistryUrl=private-repo:8116 -DimagePrefix=prefix -DimageVersion=0.5.0-2111.f48a00a -Dtier2Type=hdfs -Dtier2Config=uri=hdfs://10.XXX.XX.52:8020/ ,root=/example,replicationFactor=3```

* ECS based tier2

```gradle --info startK8SystemTests -DdockerRegistryUrl=private-repo:8116 -DimagePrefix=prefix -DimageVersion=0.5.0-2111.f48a00a -Dtier2Type=ecs -Dtier2Config=uri=http://10.XXX.XX.52:8020/ ,bucket=shared,root=example,namespace=pravega,credentials=ecs-credentials```

**How to verify it**  
It has been verified with NFS based tier 2 for Pravega. It should be verified with HDFS and ECS based tier2 for Pravega.

### CODE REVIEW

**Comment by shrids** [comment]
[Issue 3368: Ensure tier2 for Pravega is configurable when deployed by Kubernetes framework](https://trello.com/c/D6NENWGo/120-issue-3368-ensure-tier2-for-pravega-is-configurable-when-deployed-by-kubernetes-framework)

File: test/system/src/main/java/io/pravega/test/system/framework/services/kubernetes/AbstractService.java [hunk 1]
```diff
@@ -74,6 +76,10 @@
     private static final String PRAVEGA_BOOKKEEPER_VERSION = System.getProperty("pravegaBookkeeperVersion", PRAVEGA_VERSION);
     private static final String PRAVEGA_OPERATOR_VERSION = System.getProperty("pravegaOperatorVersion", "latest");
     private static final String PREFIX = System.getProperty("imagePrefix", "pravega");
+    private static final String TIER2_NFS = "nfs";
+    private static final String TIER2_ECS = "ecs";
```
Comment (fpj) [inline test/system/src/main/java/io/pravega/test/system/framework/services/kubernetes/AbstractService.java]: I'm wondering whether we really need to create these constants. For one thing, imagine that in the future we implement new bindings. We would have to remember to come back here and create a new constant. Granted that we won't do it often.

From the code below, it looks like you are only doing something special for `NFS`, so could only distinguish between `NFS` and others?

Comment (shrids) [reply test/system/src/main/java/io/pravega/test/system/framework/services/kubernetes/AbstractService.java]: yes, it is not required. Refactored the code.

**Comment by shrids** [comment]
Update regarding testing: 
This has been verified with NFS as tier2. This patch needs to be verified with HDFS and ECS as tier2, @anishakj is helping me to test this out.

**Comment by adrianmo** [comment]
To configure HDFS as Tier 2, the operator expects the following spec:

```
tier2:
  hdfs:
    uri: hdfs://10.28.2.14:8020/
    root: /example
    replicationFactor: 3
```
Note that `replicationFactor` **must be an integer**, not a string, otherwise Kubernetes will fail to parse the CR spec.

To configure ECS as Tier 2, the operator expects the following spec:

```
tier2:
  ecs:
    uri: https://object.ecstestdrive.com
    bucket: pravega-example
    root: "pravega/example"
    namespace: "131296597082943894"
    credentials: ecs-credentials
```

The above configuration works on ECS Test Drive, but it can be configured to target any other ECS system.

Prior to creating the `PravegaCluster` resource, make sure to pre-create the bucket and create a `Secret` that matches the `credentials` field and contains the access and secret key values. The following example has been tested and it is working:

```
apiVersion: v1
kind: Secret
metadata:
  name: ecs-credentials
type: Opaque
stringData:
  ACCESS_KEY_ID: QWERTY@ecstestdrive.emc.com
  SECRET_KEY: 0123456789
```

**Comment by shrids** [comment]
Update : verified that Pravega is deployed by `pravega-operator` with HDFS as tier2. Verified that `io.pravega.test.system.PravegaTest` works fine on a minikube cluster.
```
gradle --info startK8SystemTests -DdockerRegistryUrl=docker-repo:8116 -DimagePrefix=tag-DimageVersion=0.5.0-2111.f48a00a -Dtier2Type=hdfs -Dtier2Config=uri=hdfs://10.240.135.46:8020,root=/example,replicationFactor=3 --tests io.pravega.test.system.PravegaTest
```

**Comment by anishakj** [comment]
Tested with hdfs tier2.

**Pull request merged**